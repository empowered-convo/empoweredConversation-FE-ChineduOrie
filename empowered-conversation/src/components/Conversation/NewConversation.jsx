import React from "react";
import axios from "axios";
import {
  Wrapper,
  Container,
  Img,
  H3,
  FormHeader,
  Form,
  Row,
  InputGroup,
  Input,
  Bar,
  Label,
  Select
} from "./Styles";
import { Submit } from "../styled/reusables";
import Spinner from "../UI/Spinner/Spinner";
import logo from "../../img/EmpoweredConversation-Logo-Md.png";
import Modal from "../UI/Modal/Modal";

const initialForm = {
  survivorName: "",
  survivorPhone: "",
  ffName: "",
  ffPhone: "",
  categoryId: ""
};
const initialError = {
  survivorName: "",
  survivorPhone: "",
  ffName: "",
  ffPhone: "",
  categoryId: ""
};
const BASE_URL = "https://emp-convo.herokuapp.com";
class NewConversation extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      form: initialForm,
      errors: initialError,
      formValid: true,
      extractedErrors: [],
      showErrorModal: false,
      showConsentModal: false,
      categories: [],
      showSuccessModal: false,
      successMessage: "",
      isLoading: false
    };
  }
  componentDidMount() {
    axios
      .get(`${BASE_URL}/categories`)
      .then(res => {
        this.setState(prevState => ({
          ...prevState,
          categories: res.data
        }));
      })
      .catch(() => {
        return;
      });
  }
  inputChangeHandler = e => {
    const { name, value } = e.target;
    this.validateInput(name, value);
    this.setState(prevState => ({
      ...prevState,
      form: {
        ...prevState.form,
        [name]: value
      }
    }));
  };
  validateInput = (name, value) => {
    const { errors } = this.state;
    switch (name) {
      case "survivorName":
        errors.survivorName =
          value.length < 3 ? "Name cannot be less than 3 characters" : "";
        break;
      case "survivorPhone":
        errors.survivorPhone =
          value.length < 6 ? "Enter a valid phone number" : "";
        break;
      case "ffName":
        errors.ffName =
          value.length < 3 ? "Name cannot be less than 3 characters" : "";
        break;
      case "ffPhone":
        errors.ffPhone = value.length < 6 ? "Enter a valid phone number" : "";
        break;
      case "categoryId":
        errors.categoryId =
          value === "" ? "Select the conversation category" : "";
        break;
      default:
        break;
    }
    this.setState(prevState => ({
      ...prevState,
      errors
    }));
  };
  checkFormValidity = () => {
    const { form } = this.state;
    let valid = true;
    Object.values(form).forEach(val => !val.length && (valid = false));
    return valid;
  };
  onKeydown = ({ target }) => {
    const label = target.parentNode.querySelector("label");
    label.style.display = "block";
  };

  submitHandler = e => {
    e.preventDefault();
    const { errors } = this.state;
    if (!this.checkFormValidity()) {
      const extractedErrors = errors.length
        ? Object.values(errors).filter(err => err !== "")
        : ["Kindly check that all fields are filled"];
      this.setState(prevState => ({
        ...prevState,
        formValid: false,
        extractedErrors: extractedErrors,
        showErrorModal: true
      }));
    } else {
      this.setState(prevState => ({
        ...prevState,
        showConsentModal: true,
        errors: {
          ...initialError
        }
      }));
    }
  };

  toggleErrorModal = () => {
    this.setState(prevState => ({
      ...prevState,
      showErrorModal: false
    }));
  };
  toggleConsentModal = () => {
    this.setState(prevState => ({
      ...prevState,
      showConsentModal: false
    }));
  };
  submitConversation = () => {
    this.setState(prevState => ({
      ...prevState,
      showConsentModal: false,
      isLoading: true
    }));
    const { form } = this.state;
    const { survivorName, survivorPhone, ffName, ffPhone, categoryId } = form;
    const newConversation = {
      survivorname: survivorName,
      survivornumber: survivorPhone,
      ffname: ffName,
      ffnumber: ffPhone,
      category: {
        categoryid: categoryId
      }
    };
    axios
      .post(`${BASE_URL}/conversations`, newConversation)
      .then(res => {
        this.setState(prevState => ({
          ...prevState,
          successMessage: res.data.ffname,
          showSuccessModal: true,
          isLoading: false
        }));
      })
      .catch(err => {
        this.setState(prevState => ({
          ...prevState,
          extractedErrors: [err.message],
          showErrorModal: true,
          isLoading: false
        }));
      });
  };
  toggleSuccessModal = () => {
    this.setState(prevState => ({
      ...prevState,
      successMessage: "",
      showSuccessModal: false,
      form: { ...initialForm }
    }));
  };
  render() {
    const {
      form,
      errors,
      formValid,
      showErrorModal,
      extractedErrors,
      showConsentModal,
      categories,
      showSuccessModal,
      successMessage,
      isLoading
    } = this.state;
    const { survivorName, survivorPhone, ffName, ffPhone } = form;
    const {
      survivorName: sNError,
      survivorPhone: sPError,
      ffName: ffNError,
      ffPhone: ffPError,
      categoryId: catError
    } = errors;
    return (
      <Wrapper>
        {isLoading && <Spinner />}
        {!formValid && (
          <Modal
            show={showErrorModal}
            clicked={this.toggleErrorModal}
            toggle={this.toggleErrorModal}
            modalType="ok"
            modalTitle="Error"
          >
            {extractedErrors.map(err => (
              <p key={err} style={{ color: "#D62246" }}>
                {err}
              </p>
            ))}
          </Modal>
        )}
        {
          <Modal
            show={showConsentModal}
            clicked={this.submitConversation}
            toggle={this.toggleConsentModal}
            modalTitle="Confirmation"
          >
            <p>Are you sure you want to submit this conversation request?</p>
            <p>By Clicking Continue, you agree by the terms and condition</p>
          </Modal>
        }
        {
          <Modal
            show={showSuccessModal}
            toggle={this.toggleSuccessModal}
            clicked={this.toggleSuccessModal}
            modalTitle="Success"
            modalType="ok"
          >
            {successMessage && (
              <p>
                {`${successMessage}
                  has been notified, you'll be contacted when
                he/she is ready for the conversation`}
              </p>
            )}
          </Modal>
        }
        <Container>
          <FormHeader>
            <Img src={logo} alt="" />
            <H3>Start a Conversation</H3>
          </FormHeader>
          <Form autoComplete="off" onSubmit={this.submitHandler}>
            <Row>
              <InputGroup>
                <Label>Your Name</Label>
                <Input
                  type="text"
                  name="survivorName"
                  placeholder="Your Name"
                  onKeyDown={this.onKeydown}
                  onChange={this.inputChangeHandler}
                  value={survivorName}
                />
                <Bar invalid={sNError} />
              </InputGroup>
              <InputGroup>
                <Label>Your Phone Number</Label>
                <Input
                  type="tel"
                  name="survivorPhone"
                  placeholder="Your Phone Number"
                  onKeyDown={this.onKeydown}
                  onChange={this.inputChangeHandler}
                  value={survivorPhone}
                />
                <Bar invalid={sPError} />
              </InputGroup>
            </Row>
            <Row>
              <InputGroup>
                <Label>Friend/Family Name</Label>
                <Input
                  type="text"
                  name="ffName"
                  placeholder="Friend/Family Name"
                  onKeyDown={this.onKeydown}
                  onChange={this.inputChangeHandler}
                  value={ffName}
                />
                <Bar invalid={ffNError} />
              </InputGroup>
              <InputGroup>
                <Label>Friend/Family Phone Number</Label>
                <Input
                  type="tel"
                  name="ffPhone"
                  placeholder="Friend/Family Phone Number"
                  onKeyDown={this.onKeydown}
                  onChange={this.inputChangeHandler}
                  value={ffPhone}
                />
                <Bar invalid={ffPError} />
              </InputGroup>
            </Row>
            <Row>
              <InputGroup width="35%">
                <Select onChange={this.inputChangeHandler} name="categoryId">
                  <option value="">Select Category</option>
                  {categories &&
                    categories.map(cat => (
                      <option value={cat.categoryid} key={cat.categoryid}>
                        {cat.categoryname}
                      </option>
                    ))}
                </Select>
                <Bar invalid={catError} />
              </InputGroup>
            </Row>
            <Row>
              <Submit type="submit">Continue</Submit>
            </Row>
          </Form>
        </Container>
      </Wrapper>
    );
  }
}

export default NewConversation;
