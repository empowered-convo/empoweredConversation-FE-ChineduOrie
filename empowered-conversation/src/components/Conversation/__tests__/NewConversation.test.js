import { shallow } from "enzyme";
import React from "react";
import NewConversation from "../NewConversation";

describe("Conversation Component tests", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<NewConversation />);
  });
  it("should render without crashing", () => {
    expect(wrapper).toMatchSnapshot();
  });
});
