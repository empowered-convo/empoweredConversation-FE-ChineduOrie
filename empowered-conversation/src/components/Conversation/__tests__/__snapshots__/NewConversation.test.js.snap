// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversation Component tests should render without crashing 1`] = `
<styled.div>
  <Modal
    clicked={[Function]}
    modalTitle="Confirmation"
    modalType="confirm"
    show={false}
    toggle={[Function]}
  >
    <p>
      Are you sure you want to submit this conversation request?
    </p>
    <p>
      By Clicking Continue, you agree by the
      <span
        style={
          Object {
            "color": "#24b4a5",
            "cursor": "pointer",
            "textDecoration": "underline",
          }
        }
      >
         terms and condition
      </span>
    </p>
  </Modal>
  <Modal
    clicked={[Function]}
    modalTitle="Success"
    modalType="ok"
    show={false}
    toggle={[Function]}
  />
  <styled.div>
    <styled.div>
      <styled.img
        alt=""
        src="EmpoweredConversation-Logo-Md.png"
      />
      <styled.h3>
        Start a Conversation
      </styled.h3>
    </styled.div>
    <styled.form
      autoComplete="off"
      onSubmit={[Function]}
    >
      <styled.div>
        <styled.div
          invalid=""
        >
          <styled.label>
            Your Name
          </styled.label>
          <styled.input
            name="survivorName"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Your Name"
            type="text"
            value=""
          />
          <styled.span />
        </styled.div>
        <styled.div
          invalid=""
        >
          <styled.label>
            Your Phone Number
          </styled.label>
          <styled.input
            name="survivorPhone"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Your Phone Number"
            type="tel"
            value=""
          />
          <styled.span />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div
          invalid=""
        >
          <styled.label>
            Friend/Family Name
          </styled.label>
          <styled.input
            name="ffName"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Friend/Family Name"
            type="text"
            value=""
          />
          <styled.span />
        </styled.div>
        <styled.div
          invalid=""
        >
          <styled.label>
            Friend/Family Phone Number
          </styled.label>
          <styled.input
            name="ffPhone"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Friend/Family Phone Number"
            type="tel"
            value=""
          />
          <styled.span />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.div
          width="35%"
        >
          <styled.select
            name="categoryId"
            onChange={[Function]}
          >
            <option
              value=""
            >
              Select Category
            </option>
          </styled.select>
          <styled.span
            invalid=""
          />
        </styled.div>
      </styled.div>
      <styled.div>
        <styled.button
          type="submit"
        >
          Continue
        </styled.button>
      </styled.div>
    </styled.form>
  </styled.div>
</styled.div>
`;
